core/frontend/src/test/render/webgl/System.test.ts

```diff
@@ -118,4 +124,108 @@ describe("RenderSystem", () => {
     const sys2 = System.create({ contextAttributes: sys2Attrs });
     expectAttributes(sys2, { ...sys2Attrs, powerPreference: "high-performance", premultipliedAlpha: true });
   });
+
+  describe("createTextureFromImageSource", () => {
+    const imodel = {} as unknown as IModelConnection;
+
+    // This is an encoded png containing a 3x3 square with white in top left pixel, blue in middle pixel, and green in
+    // bottom right pixel.  The rest of the square is red.
+    const imageSource = new ImageSource(new Uint8Array([
+      137, 80, 78, 71, 13, 10, 26, 10, 0, 0, 0, 13, 73, 72, 68, 82, 0, 0, 0, 3, 0, 0, 0, 3, 8, 2, 0, 0, 0, 217, 74, 34,
+      232, 0, 0, 0, 1, 115, 82, 71, 66, 0, 174, 206, 28, 233, 0, 0, 0, 4, 103, 65, 77, 65, 0, 0, 177, 143, 11, 252, 97, 5, 0, 0,
+      0, 9, 112, 72, 89, 115, 0, 0, 14, 195, 0, 0, 14, 195, 1, 199, 111, 168, 100, 0, 0, 0, 24, 73, 68, 65, 84, 24, 87, 99, 248, 15,
+      4, 12, 12, 64, 4, 198, 64, 46, 132, 5, 162, 254, 51, 0, 0, 195, 90, 10, 246, 127, 175, 154, 145, 0, 0, 0, 0, 73, 69, 78, 68, 174, 66, 96, 130,
+    ]), ImageSourceFormat.Png);
+
+    before(async () => {
+      await IModelApp.startup();
+    });
+
+    after(async () => {
+      await IModelApp.shutdown();
+    });
+
+    function requestTexture(key: string | undefined, source?: ImageSource): Promise<RenderTexture | undefined> { // eslint:disable-line @typescript-eslint/promise-function-async
+      const params = new RenderTexture.Params(key, RenderTexture.Type.Normal);
+      return IModelApp.renderSystem.createTextureFromImageSource(source ?? imageSource, imodel, params);
+    }
+
+    it("should return the same Promise for the same texture", async () => {
+      const p1 = requestTexture("a");
+      const p2 = requestTexture("a");
+      expect(p1).to.equal(p2);

```

Comment:
@wgoehrig would you mind taking a look at this? I want to test something seemingly simple: if I return the same Promise from repeated calls to a function, I should get the same Promise back. But slapping `async` on a function triggers magic that wraps the returned Promise in another Promise. So I had to remove `async` all the way down the call chain, and tell eslint to stop complaining at me. Is there a better way to do this?

Reply:
@wgoehrig any thoughts? Or is this "good enough"?

Reply:
Hmm... this is a tricky one... is it not enough to just assert that both promises resolve to the same object?

If not, I guess what you have isn't _that_ evil... 😄

Reply:
No; they would have resolved to the same object previously too.

Reply:
I guess I don't understand - why does it need to be the exact same promise then?  Why do we care if the promises are wrapped?

Reply:
The bug is:
- We download two or more tiles containing the same texture image.
- We ask the render system to give us a texture object from the image.
- The render system hasn't seen this texture yet (look up by Id in cache => nothing) so we must create the texture object.
- We decode the texture image once per tile that has requested the texture object.
- 
These changes prevent us from doing the multiple decodes. That can cause spikes in memory usage and delays in loading tiles, especially with big texture images.
The only reason I care about comparing the promises for equality is so that my unit test can verify the fix and catch future regressions.

Reply:
I suppose the tests could monkey-patch System.createTextureFromImageSource() to detect requests to decode an image that already has a pending request.

Reply:
Thanks for the input, @wgoehrig. I restored `async` and adjusted the tests.